### PR TITLE
Collection level object locking.

### PR DESCRIPTION
**Pull request by MorganDawe**
Allows for collection level default object lock length.

### CODE REVIEW

File: islandora_object_lock.module [hunk 1]
```diff
@@ -45,6 +45,82 @@ function islandora_object_lock_menu() {
...
 /**
+ * Implements hook_islandora_basic_collection_build_manage_object().
+ */
+function islandora_object_lock_islandora_basic_collection_build_manage_object($form_state, $object) {
+  $return_form['manage_collection_object']['manage_obj_lock'] = array(
+    '#id' => 'manage-obj-lock',
+    '#group' => 'manage_obj_object',
+    '#access' => TRUE,
+    '#type' => 'fieldset',
+    '#title' => t('Manage Lock Objects'),
```
Comment (mitchmac) [inline islandora_object_lock.module]: Be nice to keep the case consistent with the other parts of the form.

File: islandora_object_lock.module [hunk 2]
```diff
@@ -45,6 +45,82 @@ function islandora_object_lock_menu() {
 }
 
 /**
+ * Implements hook_islandora_basic_collection_build_manage_object().
+ */
+function islandora_object_lock_islandora_basic_collection_build_manage_object($form_state, $object) {
+  $return_form['manage_collection_object']['manage_obj_lock'] = array(
+    '#id' => 'manage-obj-lock',
+    '#group' => 'manage_obj_object',
+    '#access' => TRUE,
```
Comment (mitchmac) [inline islandora_object_lock.module]: Check perms

File: islandora_object_lock.module [hunk 3]
```diff
@@ -45,6 +45,82 @@ function islandora_object_lock_menu() {
...
+ *
+ * @return array
+ *   The form to be appended to the manage object.
+ */
+function islandora_object_lock_manage_lock_form(array $form, array &$form_state, AbstractObject $object) {
+  $form_state['storage'] = array('pid' => $object->id);
+  $global_lock_length = variable_get('islandora_object_lock_duration', 30);
+  $str = str_replace(array(':'), array('_'), $object->id);
+  return array(
+    'text_field' => array(
```
Comment (mitchmac) [inline islandora_object_lock.module]: More descriptive name?

**Comment by willtp87** [comment]
This is a novel approach https://github.com/discoverygarden/islandora_object_lock/pull/5/files#L1R113 .  A table in the db or a rel on the collection would be more easily anticipated, I would favor the table to keep these settings site specific.  Another thing to consider here is that all Drupal variables are loaded on every page load.  Not sure this is worth the overhead for N number of collections.  If you stick with the variables then they must be deleted in the module uninstall and should also be deleted if the collection object is deleted. Also, the docs say to only use lower case for variable names. A relevant discussion on dynamic Drupal variables is here: http://drupal.cocomore.com/blog/how-not-delete-drupal-variables-your-modules-uninstall-function#comment-328

**Comment by willtp87** [comment]
I'm a little concerned over the consistent use of a 'collection lock' metaphor for the collection level lock duration configuration.  This is misleading suggesting at first glance that we are putting management locks on entire collections.

File: islandora_object_lock.module [hunk 4]
```diff
@@ -197,6 +282,7 @@ function islandora_object_lock_set_object_lock($pid, $user = NULL, $duration = N
       $object->relationships->add(ISLANDORA_RELS_EXT_URI, ISLANDORA_OBJECT_LOCK_IS_USER_MANAGEMENT_RULE_PRED, "TRUE", RELS_TYPE_PLAIN_LITERAL);
     }
   }
+  watchdog('Object Locked', "Object $pid has been locked for $duration min");
```
Comment (willtp87) [inline islandora_object_lock.module]: Type is usually the module name, and $pid should be pumped through the third arg. https://api.drupal.org/api/drupal/includes!bootstrap.inc/function/watchdog/7

File: islandora_object_lock.module [hunk 5]
```diff
@@ -303,9 +388,38 @@ function islandora_object_lock_is_locked(AbstractObject $object) {
...
+      $lock = islandora_object_lock_get_collection_lock($parent_pid);
+      array_push(
+        $locks,
+        $lock->duration
+      );
+    }
+  }
+  else {
+    array_push(
+    $locks,
```
Comment (willtp87) [inline islandora_object_lock.module]: I think if this is going to be a mulitiline it needs to be indented.

File: islandora_object_lock.module [hunk 6]
```diff
@@ -289,6 +434,26 @@ function islandora_object_lock_get_lock_expiry(AbstractObject $object) {
...
 /**
+ * Gets a list of all parent pids.
+ *
+ * @param AbstractObject $object
+ *   The object.
+ *
+ * @return array
+ *   The list of parent PIDs.
+ */
+function islandora_object_lock_get_parent_pids(AbstractObject $object) {
```
Comment (willtp87) [inline islandora_object_lock.module]: We have a soft dependency on the collection solution pack, we should either make that hard or throw module exists everywhere and use https://github.com/Islandora/islandora_solution_pack_collection/blob/7.x/includes/utilities.inc#L171

File: islandora_object_lock.module (hunk 4 above)
Comment (MorganDawe) [inline islandora_object_lock.module]: Did not mean to leave that in. I will remove it.

File: islandora_object_lock.module [hunk 7]
```diff
@@ -255,9 +285,24 @@ function islandora_object_lock_access_lock(AbstractObject $object, $op = NULL) {
...
-  $duration = variable_get('islandora_object_lock_duration', 30);
-  return $duration;
+  module_load_include('inc', 'islandora_object_lock', 'includes/utilities');
+  $locks = array();
+
+  if (module_exists('islandora_object_lock_length')) {
+    module_load_include('inc', 'islandora_object_lock_length', 'includes/utilities');
+    $lock = islandora_object_lock_length_compute_collection_lock_length($pid);
+  }
+  else {
```
Comment (mitchmac) [inline islandora_object_lock.module]: The previous $lock never gets used, probably shouldn't be an else?

File: islandora_object_lock.module [hunk 8]
```diff
@@ -255,9 +285,24 @@ function islandora_object_lock_access_lock(AbstractObject $object, $op = NULL) {
  *   The duration in minutes.
  */
 function islandora_object_lock_compute_duration($pid) {
-  // @TODO: Do a collection check
-  $duration = variable_get('islandora_object_lock_duration', 30);
-  return $duration;
+  module_load_include('inc', 'islandora_object_lock', 'includes/utilities');
```
Comment (mitchmac) [inline islandora_object_lock.module]: This needed?

File: islandora_object_lock.module [hunk 9]
```diff
@@ -255,9 +285,24 @@ function islandora_object_lock_access_lock(AbstractObject $object, $op = NULL) {
...
+  $locks = array();
+
+  if (module_exists('islandora_object_lock_length')) {
+    module_load_include('inc', 'islandora_object_lock_length', 'includes/utilities');
```
Comment (mitchmac) [inline islandora_object_lock.module:295]: This needed? islandora_object_lock_length_compute_collection_lock_length is in islandora_object_lock_length.module

File: islandora_object_lock_length/islandora_object_lock_length.module [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+  $default_lock_length = variable_get('islandora_object_lock_duration', 30);
+  // Only add the lock if it has a value different then the
+  // global collection lock length.
+  if ($form['duration']['#value'] != $default_lock_length && !$collection_lock_length) {
+    islandora_object_lock_length_add_collection_lock($form_state['storage']['pid'], $form['duration']['#value']);
+  }
+  else {
+    // If this lock is the same as the global lock, remove it from
+    // the collection locks table.
+    if ($default_lock_length == $collection_lock_length->duration) {
```
Comment (mitchmac) [inline islandora_object_lock_length/islandora_object_lock_length.module]: $collection_lock_length->duration can be not set if the lock length isn't in the db.

File: islandora_object_lock_length/islandora_object_lock_length.module [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+ *   The current $form_state.
+ */
+function islandora_object_lock_length_manage_lock_form_submit(array $form, array &$form_state) {
+  module_load_include('inc', 'islandora_object_lock_length', 'includes/utilities');
+  // Get Collection lock duration if it is already set.
+  $collection_lock_length = islandora_object_lock_length_get_collection_lock($form_state['storage']['pid']);
+  $default_lock_length = variable_get('islandora_object_lock_duration', 30);
+  // Only add the lock if it has a value different then the
+  // global collection lock length.
+  if ($form['duration']['#value'] != $default_lock_length && !$collection_lock_length) {
```
Comment (mitchmac) [inline islandora_object_lock_length/islandora_object_lock_length.module]: This doesn't look right, adding a new value the first time inserts NULL in the db.

**Pull request merged**